docs/adr/0007-modelling-history.md

```diff
@@ -0,0 +1,272 @@
+revisions. A consequence of this is that an edit to a document overwrites the
+previous state of the document. Since a document can exist in two forms, draft
+and live, and Content Publisher only knows of one we hit problems when
+knowledge of both would be useful.
+
+The suggestion is that we enhance Content Publisher to support the concepts of
+Document, Edition and Revision. The features that we'd like this to enable
+are:
+
+For a document with a live version and a draft:
+
+- The ability to discard the draft and revert the current version in Content
+  Publisher back to the live version (including all images and attachments) to
+  be in sync with Publishing API;
+- The ability to have the details of the live version, most noticeably state
+  and URL so this can be presented to users;
+- Ability to re-present either the live or draft version to Publishing API
+  should there be any problems with publishing.
+
+And with regards to history:
+
+- To know who did what to a document and when - effectively an audit trail of
+  accountability;
+- Be able to compare two revisions of a document to show changes;
+- Have stored in our database a full history of a document, which may go back
+  further than the Publishing API (Whitehall has longer histories of some
+  published documents)
+
+A [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method) breakdown of this
+functionality was defined as:
+
+### Must
+
+- Store full information on latest and live versions of a document
+- Consider how a revision of a document and aspects of it have state that may
+  change
+
+### Should
+
+- Have the full information for every edition that was published on GOV.UK
+- Record every change a publisher makes and have that information accessible
+- Consider states to be something that may change in a background job
+- Avoid relying on the TimelineEntry model as the sole answer to any historical
+  queries
+
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice
+- Enable a future feature to enable reverting to different points in time
+- Separate related data to different tables to reduce data usage
+
+### Won’t
+
+- Consider the sharing of data between translations, as per
+  [ADR-3](../adr/0003-initial-domain-modelling.md#decision)
+
+## Decision
+
+The proposed data model expands on the definitions of entities accordingly:
+
+**Document**: A record that represents all versions of a piece of content in a
+particular locale. It has many editions and many revisions. At any time it will
+have a current edition -  shown on Content Publisher index - and potentially a
+live edition which is currently on GOV.UK. The live and current edition can be
+the same. Document serves as a mutable entity that tracks when a Document was
+last edited and it is expected to be a joining point for document related data
+that is not associated with a particular editions.
+
+**Edition**: An edition is a numbered version of a document that has been, or is
```

Comment:
> A few questions here:
> 
> * What does 'current revision' mean if we have a separate draft and live edition?

Each edition is associated with a particular revision that represents the content for that edition. I went with calling that `current_revision`

So a draft edition may have a current revision of 44
And the live edition may have a current revision of 32

Is the terminology of "current" tricky here? I now realise how it could be because we have a current edition on a document. 

> * Could you clarify the need to know past states of an edition outside of a timeline?

That's the main purpose, secondary usage would be auditing and debugging - which is a realistic case. It's based on the opinion that TimelineEntry should be avoided to be canonical source of information as much as possible.

> * Could you give an example of a constraint for editions?

We'd probably only want to have one edition for a document with a particular number
We may want to maintain a uniqueness constraint for base paths and draft/live stacks.

Reply:
I'm still unclear as to the difference between revisions and editions.

Reply:
Is it supposed to work like books? Where an author(publisher) can send multiple revisions to their editor(2i) for review, but these relate to one edition, and the final revision is what becomes the published edition?

Reply:
@leenagupte You've got that exactly right 👍